### ISSUE TITLE
"restore" procedure does not finish and can only be interrupted by killing GDL

### ISSUE THREAD

**Issue by brandy125** [gnudatalanguage/gdl#1545]
Up to a size of approx. 4GB or up to array sizes of 10k x 10k the save and restore procedures work fine for me.
Now I have bigger arrays so that the saved file is around 4.7GB (or arrays bigger than 10k x 10k).
When I try to restore it never finishes and I can kill it only by killing GDL.

I tried to reproduce the behaviour with a single array but did not succeed even if the saved file gets above 5 GB.

How can I debug this?

Here is what I get when the restore works:

```
strace gdl <<< "restore,'small.sav'"

read(8, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 4096) = 4096
read(8, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 4096) = 4096
read(8, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 4096) = 4096
read(8, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 4096) = 4096
```

and this I get when the restore does not work and never finishes:

```
strace gdl <<< "restore,'big.sav'"

lseek(8, 4096, SEEK_SET)                = 4096
lseek(8, 4096, SEEK_SET)                = 4096
lseek(8, 4096, SEEK_SET)                = 4096
lseek(8, 4096, SEEK_SET)                = 4096
```

Any guess?

**Comment by alaingdl** [comment]
Which system do you use ? Maybe no memory enough ?
I just tested on a __small__ server (multi-core, Linux Debian, 256 Go of RAM ...) without problem

Size is over 5 Go : 
```
$ ls -l
-rw-r--r-- 1 alaingdl alaingdl 5640262920 mars  12 15:36 a_rand_10kx10kx500.xdr
$ ls -lh
-rw-r--r-- 1 alaingdl alaingdl 5,3G mars  12 15:36 a_rand_10kx10kx500.xdr
```

 Writing and reading back are OK
```
GDL> a=randomu(seed, 10000, 10000, 50,/double)
GDL> save, file='a_rand_10kx10kx500.xdr', a   
GDL> help, a
A               DOUBLE    = Array[10000, 10000, 50]
GDL> a=0.
GDL> tic & restore, 'a_rand_10kx10kx500.xdr', /verb & toc
% Portable (XDR) SAVE/RESTORE file.
% Save file written by alaingdl@jwst1, Sun Mar 12 15:35:23 2023
% Restored variable: A.
% Time elapsed :        41.070958 seconds.
GDL> help, a
A               DOUBLE    = Array[10000, 10000, 50]
```
(time is not very indicative because many others consuming tasks are running at the same time)

**Comment by brandy125** [comment]
I am running Suse Tumbleweed inside Parallels on a Mac Ultra with 128GB RAM.
The Linux has 100GB RAM and 20 CPUs. This is not the problem.

I tried with a single array as you did and it works fine, as well with arrays much bigger than 5GB.

It is very weird.

Up to now I just can say that it constantly fails.

I have around 150 variables inside the .sav file.
I will try to reduce the variables until I find the one that is causing the problem.

**Comment by alaingdl** [comment]
In fact I did not understood you have several arrays in one save file, I will try that ASAP. It is possible the way the internal offsets are written or managed in the code, of the file, cannot manage such large jumps in the file ?

**Comment by brandy125** [comment]
I did not explain well, sorry.


Here is the beginning of the list of variables when the restore hangs:

```AGAIN           DOUBLE    = Array[9081, 9080]
ALFA            FLOAT     =      0.540000
ALT             FLOAT     = Array[9081, 9080]
ALTNAVI         FLOAT     = Array[1, 5683]
AUX_T1          LONG      = Array[5]
BAND            STRING    = 'l'
BW              DOUBLE    =    1.5000000e+08
C0              DOUBLE    =    2.9979246e+08
CH1M            INT       =        0
CHAN            INT       =        0
DATA            INT       =        0
DAY             INT       =        6
DELR            DOUBLE    =        4.9117996
DELTAX          FLOAT     =     0.0500000
DELTAY          FLOAT     =     0.0500000
DTIME           INT       =        8
FC              DOUBLE    =    1.2750000e+09
FDCJ            DOUBLE    = Array[5683]
```

And this would be the list when I start my program with different parameters so that the restore will work:

```
AGAIN           DOUBLE    = Array[8001, 8000]
ALFA            FLOAT     =      0.540000
ALT             FLOAT     = Array[8001, 8000]
ALTNAVI         FLOAT     = Array[1, 5683]
AUX_T1          LONG      = Array[5]
BAND            STRING    = 'l'
BW              DOUBLE    =    1.5000000e+08
C0              DOUBLE    =    2.9979246e+08
CH1M            INT       =        0
CHAN            INT       =        0
DATA            INT       =        0
DAY             INT       =        6
DELR            DOUBLE    =        4.9117996
DELTAX          FLOAT     =     0.0500000
DELTAY          FLOAT     =     0.0500000
DTIME           INT       =        8
FC              DOUBLE    =    1.2750000e+09
FDCJ            DOUBLE    = Array[5683]
```

**Comment by brandy125** [comment]
Here is how to reproduce the error:

```

a       =fltarr(10001, 10000)
b       =dblarr(10001, 10000)
c       =complexarr(10001, 10000)
d       =complexarr(3200, 5683)
e       =fltarr(10001, 10000)
f       =dblarr(5683)
g       =dblarr(255604)
h       =dblarr(5683)
i       =dblarr(255604)
j       =dblarr(5683)
k       =dblarr(255604)
l       =dblarr(10001, 10000)
m       =dblarr(10001, 10000)
n       =dblarr(255604)
o       =dblarr(255604)
p       =dblarr(5683)
q       =dblarr(10001, 10000)
r       =dblarr(255604)
s       =dblarr(10001, 10000)
t       =dblarr(255604)
save,file='doesnotrestore.sav'
```

Then open other GDL and try to restore:

```
restore,'doesnotrestore.sav'
```

The CPU load for the restore session stays at 100%, consuming almost no memory and does never finish.

**Comment by alaingdl** [comment]
Thanks. I used that and was able to reproduce your case. I also generate the output with IDL and was not able to read it back in GDL. 

If I remember well @GillesDuvert was the main contributor to XDR support, I hope he will be able to help us.

Compiling now with the internal flag set to 1

**Comment by alaingdl** [comment]
OK, switching on the debug flag, the problem is clear : we need an offset, the offset goes above the limits and go back to zero (as we say in french : CQFD). The code is not that clear for me since we have two similar loops in restore,
both try first with fseek() which may be only 32b

On the example provided above ...
```
Offset 4154203152: record type VARIABLE
Offset 4156248084: record type VARIABLE
Offset 4156293648: record type VARIABLE
Offset 0: record type Erreur de segmentation
```

**Comment by GillesDuvert** [comment]
Hi folks,
Obviously there is a problem of seeking an address that is larger than the pointer type used locally, such 'details' are lurking in all the GDL code I'm afraid.
I suspect IDL to embed in its save file a code saying that the xdr pointers must  be 64 bits, and of course GDL falled in the trap.
This may be quite simple, or not at all... We'll see.
(Note: actually I'm not a contributor to XDR, which is an external library, but I indeed wrote the save/restore procedure)

**Comment by GillesDuvert** [comment]
solved --- but see #1550
